### DISCUSSION TITLE
Workaround for installing Linux on Surface Pro where unable to boot (testet on gen7)

### DISCUSSION THREAD

**Discussion by ThorsFarm** [linux-surface/linux-surface#1455] [General]
_Updated with help from @Raeitus and fresh install on Fedora 41 with updates for DNF5_
I have no solution, but I have a work-around I made last night.
I use Fedora, but I guess it would work for all distro that have a non-bootable kernel.
`<disclaimer>This is a summary of many trial and error, so there could be wrong information</disclaimer> 
`
1. Use Ventoy with MBR, NTFS and no Secure Boot as the boot media
2. Disable Secure Boot on surface
3. Boot from Ventoy, make sure you choose grub2 at the menu!
4. Install Fedora (this should be normal click-click-next)
5. After installment on Surface, either dont reboot or boot from live-cd once again, the surface will not boot with the stock Fedora kernel.
6. Make sure you have internet connection!

Run the following commands from a terminal:
```
sudo chroot /mnt/sysimage/
```
For Fedora <41 with DNF4:
```
dnf config-manager --add-repo=https://pkg.surfacelinux.com/fedora/linux-surface.repo
dnf install --allowerasing -y kernel-surface iptsd libwacom-surface
dnf reinstall kernel-surface*
grub2-mkconfig -o /boot/grub2/grub.cfg
```
For Fedora 41 with DNF5
```
dnf config-manager addrepo --from-repofile=https://pkg.surfacelinux.com/fedora/linux-surface.repo
dnf --disablerepo=updates -y install --allowerasing kernel-surface iptsd libwacom-surface
```
If you have secure boot enabled:
```
dnf -y install surface-secureboot
```
After reboot you should boot into enrollment of keys, the password is: `surface`

**Comment by ChromiteExabyte** [comment]
## Replicating Fedora -> Surface Book 2 Installation Problems 

This thread helped me and I am adding to it. 

My configuration: 

- Fedora 40, KDE Spin (Plasma 6)
- UEFI Version 394.779.768
- Balena Etcher 
- The Surface Book 2's UEFI was set to have secure boot and the TPM module disabled. 
- - Microsoft has some good documentation, check there if you don't know how to get into your UEFI/ BIOS / computer-booter. 


💻 Selecting 'Launch from USB' from the UEFI _should_ allow a normal boot where you can select "Launch Fedora 40" normally.
❌ Attempting to launch from the 'grub' boot screen would cause a stall. On that launch / stall condition:
-> There is a '_' char in the upper left corner of the display. 
-> There is a white keyboard .svg icon overlay in the bottom right corner of the display. 
-> This state is persistent and keyboard / touch inputs do not affect screen state. 

✔️ Using the Ventoy tool, 
-> Ventoy is open source and pretty nifty. 
-> An old 8GB USB drive was formatted through the Ventoy tool following the instructions on the website 

The Ventoy tool works ! There is a blue/gray GUI and options to configure the launch. The .iso image, Fedora 40, is visible in the Ventoy launcher. (I forgot to do this at first, and there was no option to boot from an image that didn't exist. Drag the .iso to the Ventoy formatted drive right after you format it and before you boot). 

❕ GRUB-2 is something that can be selected for launch through Ventoy. Do this prior to launching from Ventoy, as stated above. 
❔ I don't know if not doing the above step will cause a different outcome here. 

From this point onward, the Fedora 40 grand-unified-bootloader (grub) will launch successfully. The text will still be tiny as it launches. 

👍 This will get a person into a Fedora 40 desktop following the above steps. 

I don't know if the Surface Book 2 will have the same difficulties on rebooting. 
A temporary solution is to keep the system running continuously.

**Comment by ghost** [comment]
Instead of all of those mounts, a much simpler thing to do is just: `sudo chroot /mnt/sysimage`

You can follow these steps: https://github.com/linux-surface/linux-surface/issues/1162#issuecomment-2005363664

* Use Ventoy to boot into Fedora Live USB
  * Some might need to select grub2 in Ventoy when booting into the Live USB/ISO
* Connect to WiFi/Internet
* Install Fedora
* Without rebooting, `chroot` into your installation: `sudo chroot /mnt/sysimage`
* Add linux-surface repos: `dnf config-manager --add-repo=https://pkg.surfacelinux.com/fedora/linux-surface.repo`
  * For DNF5 (Fedora 41+), use this instead: 
    * `dnf config-manager addrepo --from-repofile=https://pkg.surfacelinux.com/fedora/linux-surface.repo`
* Install linux-surface kernel:  `dnf --disablerepo=updates install --allowerasing kernel-surface iptsd libwacom-surface`
* Enroll secure boot keys: `dnf install surface-secureboot`
* Reboot which should begin Mokmanager prompt, enroll keys (password `surface`), and then boot successfully.

This worked for Fedora 40 without any issues for me on a Surface Pro 7

Edit: Added DNF5

**Reply by ThorsFarm** [reply]
I agree, this seems like a much cleaner install. I will try this as soon as I screw up my current install.
Thanks for contributing to the solution :)

**Reply by singh-industries** [reply]
Worked on my Surface 6 although omitted iptsd (have the ghost touch screen issue)

**Reply by tylerstewartt** [reply]
Worked for me as of 7/18/2024. I'm on a Surface Pro 7 running Fedora 40. Only thing I could find that doesn't work is the camera which isn't a deal breaker for me. I'm mainly using it as a tablet so it's nice that the touchscreen controls work great.

**Reply by Deerhall** [reply]
I had issues getting Fedora to install on my Surface Laptop 3 ever since that firmware update. This short little guide worked perfectly for me!

**Reply by sherlukhoms** [reply]
Total Linux noob here but willing to learn:
How do I `chroot` into my installation? It worked until `dnf --disablerepo=updates install --allowerasing kernel-surface iptsd libwacom-surface` , after that I got `Error: Unable to find a match: kernal-surface`

Is there something specific I need to do when doing the chroot?

**Reply by ThorsFarm** [reply]
You should chroot like this: `sudo chroot /mnt/sysimage`
It seems like you are missing the repo for linux-surface. Are you sure you have added it?
This is the command for that: `dnf config-manager --add-repo=https://pkg.surfacelinux.com/fedora/linux-surface.repo`

**Reply by sherlukhoms** [reply]
Sorry if this seems like a silly question but is there any step I have to do prior `sudo chroot /mnt/sysimage` ?
After restarting again into live version from USB I get `chroot: cannot change root directory to '/mnt/chroot' : No such file or directory`  :(

EDIT: I reinstalled Fedora Workstation without reboot (had to reboot because it didnt connect to Wifi again after being in standby over night). After that it all worked. Thanks! :)

**Comment by otjamin** [comment]
EDIT: If you did *not* reboot after the installation, you should be able to just follow https://github.com/linux-surface/linux-surface/discussions/1455#discussioncomment-9790505 instead.

I will add my two cents here. I don't think anyone has mentioned disk encryption yet and I had problems with the normal Ventoy installation with the latest firmware. Hope this helps.

### How-to (tested on SP8 with new firmware and disk encryption)

1. Download Ventoy.
2. (optional?) When installing Ventoy on the USB Stick, ~choose the GPT partition table instead of MBR~. Disable secure boot support.[^1]
4. Copy the ISO to the Ventoy drive.
5. Turn off secureboot on your surface.
6. Boot with the GRUB2 option in Ventoy.
7. Install Fedora as normal.
8. Do **not** reboot!
9. Mount the installed system.[^2]
```
$ mkdir /mnt/sysimage
$ mount -o subvol=root /dev/nvme0n1p3 /mnt/sysimage
$ mount -o subvol=home /dev/nvme0n1p3 /mnt/sysimage/home
$ mount /dev/nvme0n1p2 /mnt/sysimage/boot
$ mount /dev/nvme0n1p1 /mnt/sysimage/boot/efi
$ chroot /mnt/sysimage
```
10. Follow the surface-linux installation instructions (e.g. installing the custom kernel).[^3]
11. Reboot now and party :partying_face: 


[^1]: I do not know if this is necessary for all Surface devices. But on my SP8 it was required after a firmware upgrade.
[^2]: When using disk encryption you need to use the already mounted `/dev/mapper/luks...` instead of `/dev/nvme0n1p3`. If there is no such device you need to unlock it yourself with `cryptsetup luksOpen /dev/nvme0n1p3 mydrive`.
[^3]: Add the kernel drivers at this step if you are using [disk encryption](https://github.com/linux-surface/linux-surface/wiki/Disk-Encryption).

**Reply by ghost** [reply]
You don't need to mount all of those. Just type `sudo chroot /mnt/sysimage` after you install without rebooting.

I haven't had the need to turn off secure boot myself, but I've heard some people needed to.

I've installed Fedora on both a SP7 and SP8 following [my steps above](https://github.com/linux-surface/linux-surface/discussions/1455#discussioncomment-9790505) and all I did was select Encrypt during the partitioning and installed Fedora.

**Reply by otjamin** [reply]
Ok, I guess if you don't reboot after installation, this is still mounted.

But if you do reboot, or something else happens, this is required. As you can see https://github.com/linux-surface/linux-surface/discussions/1455#discussioncomment-10903964 and also from my own experience.

I will edit the comment accordingly.

**Comment by ruslanbay** [comment]
is this problem still present in the latest fedora 42 rawhide?

**Comment by brendankoja** [comment]
> _Updated with help from @Raeitus and fresh install on Fedora 41 with updates for DNF5_ I have no solution, but I have a work-around I made last night. I use Fedora, but I guess it would work for all distro that have a non-bootable kernel. `<disclaimer>This is a summary of many trial and error, so there could be wrong information</disclaimer> `
> 
>     1. Use Ventoy with MBR, NTFS and no Secure Boot as the boot media
> 
>     2. Disable Secure Boot on surface
> 
>     3. Boot from Ventoy, make sure you choose grub2 at the menu!
> 
>     4. Install Fedora (this should be normal click-click-next)
> 
>     5. After installment on Surface, either dont reboot or boot from live-cd once again, the surface will not boot with the stock Fedora kernel.
> 
>     6. Make sure you have internet connection!
> 
> 
> Run the following commands from a terminal:
> 
> ```
> sudo chroot /mnt/sysimage/
> ```
> 
> For Fedora <41 with DNF4:
> 
> ```
> dnf config-manager --add-repo=https://pkg.surfacelinux.com/fedora/linux-surface.repo
> dnf install --allowerasing -y kernel-surface iptsd libwacom-surface
> dnf reinstall kernel-surface*
> grub2-mkconfig -o /boot/grub2/grub.cfg
> ```
> 
> For Fedora 41 with DNF5
> 
> ```
> dnf config-manager addrepo --from-repofile=https://pkg.surfacelinux.com/fedora/linux-surface.repo
> dnf --disablerepo=updates -y install --allowerasing kernel-surface iptsd libwacom-surface
> ```
> 
> If you have secure boot enabled:
> 
> ```
> dnf install surface-secureboot
> ```
> 
> After reboot you should boot into enrollment of keys, the password is: `surface`

after banging my head against a wall for 3 hours I found this and you are a godsend thank you

**Comment by ruslanbay** [comment]
I found another workaround. One problem is that the [Linux kernel (wrongly) indicates compatibility with NX mode to GRUB](https://github.com/linux-surface/linux-surface/issues/1162). So you can just use systemd-boot instead of Grub:
https://fedoraproject.org/wiki/Changes/cleanup_systemd_install#How_To_Test

1. Boot from Fedora installation image. It's gonna look like this:
![image](https://github.com/user-attachments/assets/ca49ea7a-8f7d-4240-b5e5-81cee314830d)
2. Press e
3. Pass inst.sdboot on the kernel/grub command line presented on the install media and install as normal
4. Use the machine as normal

**Comment by 3nd0y** [comment]
> _Updated with help from @Raeitus and fresh install on Fedora 41 with updates for DNF5_ I have no solution, but I have a work-around I made last night. I use Fedora, but I guess it would work for all distro that have a non-bootable kernel. `<disclaimer>This is a summary of many trial and error, so there could be wrong information</disclaimer> `
> 
> 1. Use Ventoy with MBR, NTFS and no Secure Boot as the boot media
> 2. Disable Secure Boot on surface
> 3. Boot from Ventoy, make sure you choose grub2 at the menu!
> 4. Install Fedora (this should be normal click-click-next)
> 5. After installment on Surface, either dont reboot or boot from live-cd once again, the surface will not boot with the stock Fedora kernel.
> 6. Make sure you have internet connection!
> 
> Run the following commands from a terminal:
> 
> ```
> sudo chroot /mnt/sysimage/
> ```
> 
> For Fedora <41 with DNF4:
> 
> ```
> dnf config-manager --add-repo=https://pkg.surfacelinux.com/fedora/linux-surface.repo
> dnf install --allowerasing -y kernel-surface iptsd libwacom-surface
> dnf reinstall kernel-surface*
> grub2-mkconfig -o /boot/grub2/grub.cfg
> ```
> 
> For Fedora 41 with DNF5
> 
> ```
> dnf config-manager addrepo --from-repofile=https://pkg.surfacelinux.com/fedora/linux-surface.repo
> dnf --disablerepo=updates -y install --allowerasing kernel-surface iptsd libwacom-surface
> ```
> 
> If you have secure boot enabled:
> 
> ```
> dnf install surface-secureboot
> ```
> 
> After reboot you should boot into enrollment of keys, the password is: `surface`

I have finished installed fedora then continue update kernel use surface kernel as you all suggested, but after boot normal then continue create account and password I am stuck in the gnome desktop can't restart or access application menu as picture below. Then i tried to hard power off by long press the button power, then tried to restart to fedora again. Then i stuck in the login, even i input correct password, it will loop again ask password. Any ideas what happen here?
<img width="622" alt="image" src="https://github.com/user-attachments/assets/e3bc34d6-1eff-460c-b324-02b5a6db55a8" />

**Reply by busywait** [reply]
Just a note that I needed to follow this guide for installing Fedora 42 Workstation. Thanks for writing it!

I installed Ventoy onto my installer USB drive with NTFS, then exFAT and finally FAT32. They all allowed the Fedora installer to start, but when the USB drive was formatted with NTFS the install was *very* slow. So slow it didn't finish in hours and I started again. Setting up Ventoy on the USB stick with either FAT32 or exFAT (the default) both allowed the install to run quickly with no problem.

**Comment by Mr-KayZ** [comment]
> _Updated with help from @Raeitus and fresh install on Fedora 41 with updates for DNF5_ I have no solution, but I have a work-around I made last night. I use Fedora, but I guess it would work for all distro that have a non-bootable kernel. `<disclaimer>This is a summary of many trial and error, so there could be wrong information</disclaimer> `
> 
>     1. Use Ventoy with MBR, NTFS and no Secure Boot as the boot media
> 
>     2. Disable Secure Boot on surface
> 
>     3. Boot from Ventoy, make sure you choose grub2 at the menu!
> 
>     4. Install Fedora (this should be normal click-click-next)
> 
>     5. After installment on Surface, either dont reboot or boot from live-cd once again, the surface will not boot with the stock Fedora kernel.
> 
>     6. Make sure you have internet connection!
> 
> 
> Run the following commands from a terminal:
> 
> ```
> sudo chroot /mnt/sysimage/
> ```
> 
> For Fedora <41 with DNF4:
> 
> ```
> dnf config-manager --add-repo=https://pkg.surfacelinux.com/fedora/linux-surface.repo
> dnf install --allowerasing -y kernel-surface iptsd libwacom-surface
> dnf reinstall kernel-surface*
> grub2-mkconfig -o /boot/grub2/grub.cfg
> ```
> 
> For Fedora 41 with DNF5
> 
> ```
> dnf config-manager addrepo --from-repofile=https://pkg.surfacelinux.com/fedora/linux-surface.repo
> dnf --disablerepo=updates -y install --allowerasing kernel-surface iptsd libwacom-surface
> ```
> 
> If you have secure boot enabled:
> 
> ```
> dnf -y install surface-secureboot
> ```
> 
> After reboot you should boot into enrollment of keys, the password is: `surface`

I would also like to add the following for the Surfae Laptop Go 1 (SLG1, maybe potentially others too?) installing Fedora 42: While as CHROOT, still update the bootloader after following DNF5 commands:
```
grub2-mkconfig -o /boot/grub2/grub.cfg
```

This is due to me encountering the same grub failing to load boot commands even after installing the Surface Kernel. After numerous attempted fixes, updating grub did the trick.

Again, amazing guide and thank you sincerely for writing this! Everything else worked like a charm!

**Comment by krixoxxxx** [comment]
I had a different solution
sudo dd if=Fedora-Workstation-Live-43-1.6.x86_64.iso  of=/dev/sda bs=4M status=progress oflag=sync
i just wrote it directly to my drive (ventoy wouldnt work)
this is on a surface pro 4 btw
